### ISSUE TITLE
Removed limit in Docker Engine 23 breaks existing config file

### ISSUE THREAD

**Issue by kamulos** [haproxy/haproxy#2043]
### Detailed Description of the Problem

So this might be perfectly intentional, but I wanted to raise awareness, that this broke our system:

On Arch Linux the Docker Engine got the update to version 23.0.1. In this the nofile hard limit is now 1073741816. Because haproxy is allocating resources based on the hard limit I get the following error message when I start the haproxy now in the docker container:

```
[NOTICE]   (1) : haproxy version is 2.8-dev3-e74d77b
[NOTICE]   (1) : path to executable is /usr/local/sbin/haproxy
[ALERT]    (1) : Not enough memory to allocate 1073741815 entries for fdtab!
[ALERT]    (1) : No polling mechanism available.
  This may happen when using thread-groups with old pollers (poll/select), or
  it is possible that haproxy was built with TARGET=generic and that FD_SETSIZE
  is too low on this platform to support maxconn and the number of listeners
  and servers. You should rebuild haproxy specifying your system using TARGET=
  in order to support other polling systems (poll, epoll, kqueue) or reduce the
  global maxconn setting to accommodate the system's limitation. For reference,
  FD_SETSIZE=1024 on this system, global.maxconn=536870885 resulting in a maximum of
  1073741815 file descriptors. You should thus reduce global.maxconn by 536870396. Also,
  check build settings using 'haproxy -vv'.
```

### Expected Behavior

I would expect haproxy to choose reasonable limits even if the nofile limit is really big. This issue might be frustrating for anyone who stumbles on it.

But obviously this can easily be fixed by setting a `maxconn` value, which could be reasonable anyway.

### Steps to Reproduce the Behavior

1. Use archlinux or any System, that uses an unchanged Docker Engine 23.0.1
2. create a haproxy container with an empty config file
3. start the container

### Do you have any idea what may have caused this?

Maybe this: https://github.com/containerd/containerd/commit/c691c36614622b205a859ae7e656badb4a553076

### What is your configuration?

```haproxy
empty config file in docker container
```

### Output of `haproxy -vv`

```plain
Our system currently uses haproxy 2.4, but this also occurs with haproxy 2.8-dev


HAProxy version 2.8-dev3-e74d77b 2023/02/04 - https://haproxy.org/
Status: development branch - not safe for use in production.
Known bugs: https://github.com/haproxy/haproxy/issues?q=is:issue+is:open
Running on: Linux 6.1.11-arch1-1 #1 SMP PREEMPT_DYNAMIC Thu, 09 Feb 2023 20:06:08 +0000 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -Wall -Wextra -Wundef -Wdeclaration-after-statement -Wfatal-errors -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wno-string-plus-int -Wno-atomic-alignment
  OPTIONS = USE_GETADDRINFO=1 USE_OPENSSL=1 USE_LUA=1 USE_PROMEX=1 USE_PCRE2=1 USE_PCRE2_JIT=1
  DEBUG   = -DDEBUG_STRICT -DDEBUG_MEMORY_POOLS

Feature list : -51DEGREES +ACCEPT4 +BACKTRACE -CLOSEFROM +CPU_AFFINITY +CRYPT_H -DEVICEATLAS +DL -ENGINE +EPOLL -EVPORTS +GETADDRINFO -KQUEUE -LIBATOMIC +LIBCRYPT +LINUX_SPLICE +LINUX_TPROXY +LUA +MATH -MEMORY_PROFILING +NETFILTER +NS -OBSOLETE_LINKER +OPENSSL -OPENSSL_WOLFSSL -OT -PCRE +PCRE2 +PCRE2_JIT -PCRE_JIT +POLL +PRCTL -PROCCTL +PROMEX -PTHREAD_EMULATION -QUIC +RT +SHM_OPEN +SLZ +SSL -STATIC_PCRE -STATIC_PCRE2 -SYSTEMD +TFO +THREAD +THREAD_DUMP +TPROXY -WURFL -ZLIB

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_TGROUPS=16, MAX_THREADS=256, default=8).
Built with OpenSSL version : OpenSSL 1.1.1n  15 Mar 2022
Running on OpenSSL version : OpenSSL 1.1.1n  15 Mar 2022
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.3
Built with the Prometheus exporter as a service
Built with network namespace support.
Support for malloc_trim() is enabled.
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.36 2020-12-04
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 10.2.1 20210110

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
         h2 : mode=HTTP  side=FE|BE  mux=H2    flags=HTX|HOL_RISK|NO_UPG
       fcgi : mode=HTTP  side=BE     mux=FCGI  flags=HTX|HOL_RISK|NO_UPG
  <default> : mode=HTTP  side=FE|BE  mux=H1    flags=HTX
         h1 : mode=HTTP  side=FE|BE  mux=H1    flags=HTX|NO_UPG
  <default> : mode=TCP   side=FE|BE  mux=PASS  flags=
       none : mode=TCP   side=FE|BE  mux=PASS  flags=NO_UPG

Available services : prometheus-exporter
Available filters :
	[BWLIM] bwlim-in
	[BWLIM] bwlim-out
	[CACHE] cache
	[COMP] compression
	[FCGI] fcgi-app
	[SPOE] spoe
	[TRACE] trace
```

**Comment by TimWolla** [comment]
I believe this effectively is a duplicate of #1622.

**Comment by kamulos** [comment]
So I think the new information here would be that there is a new default in Docker and most likely the number of people running into this behavior will increase

**Comment by lukastribus** [comment]
The TL;DR is:

- set [global maxconn](http://docs.haproxy.org/2.6/configuration.html#3.2-maxconn) which simply sets the maximum amount of total connections (front and backend)
- or if you must, use haproxy 2.6+ and configure the new global parameter [fd-hard-limit](http://docs.haproxy.org/2.6/configuration.html#3.1-fd-hard-limit), which is setting the upper limit of FDs


Can't say that I like `fd-hard-limit`. To comprehend this requirement, users need understanding of ulimit and FDs as well as understanding the fallback mechanisms in haproxy that are executed, when maxconn is not configured. I would've preferred a `maxconn-hard-limit` instead.

In this way we are "educating" users that start playing with haproxy to focus on FDs as opposed to maxconn.

**Comment by wtarreau** [comment]
Initially I thought "ah no, we still have that to do" and I didn't remember I had finally done it. I don't like it either because it's just asking haproxy to do its best within some limits. I personally wouldn't like to see my LBs work like this in the dark, but I have to admit that for tests or development it can be particularly convenient (especially since it will not complain about not being able to raise a value). But yes, for production, you should know how many FDs you want to assign and set them using ulimit-n or using maxconn.

**Comment by wtarreau** [comment]
By the way, to give a bit more backgroud, we used to have a 2k maxconn limit for a very long time and it was causing much more harm than such an error: the process used to start well and was working perfectly fine until the day there was a big rush on the site and it wouldn't accept more connections than the default limit. I'm not that much tempted by setting new high default limits. We do have some users running with 2+ million concurrent connections, or roughly 5M FDs. That's already way above what most users would consider an acceptable default limit, and anything below this could mean that such users wouldn't know about the setting and could get trapped.

Maybe among the approaches to think about, these ones should be considered:
- if the process starts unprivileged and in foreground without the quiet setting, we consider it's being used by a developer and we could significantly lower the value to be nice ;
- if we recognize a large ulimit-n at boot and there's no maxconn, we could limit it and emit a warning indicating that it got limited and that an explicit maxconn needs to be used for such large values. But we do know that many users don't read warnings anymore...
- ot stay in the current situation but improve the message to detect that it's not a poller issue but really too large an FD value, and refuse to start with such large values and explain how to deal with it (and the impacts)

**Comment by kamulos** [comment]
That sounds absolutely reasonable :+1: 

Personally I like your third option a lot. We were most worried about it breaking without any obvious cause and a rather strange message. A good explanation would have saved us quite a bit of digging around.

**Comment by lukastribus** [comment]
> * ot stay in the current situation but improve the message to detect that it's not a poller issue but really too large an FD value, and refuse to start with such large values and explain how to deal with it (and the impacts)

Just to reiterate again: we should explain maxconn first, FD limits after. FD limits are the low level cousin of maxconn in some ways, it's important that we direct the novice user to high level configuration directives as opposed to the low level ones.

**Comment by wtarreau** [comment]
I generally agree with your approach, Lukas, but here it's the FD limit that implies a maxconn value and the user didn't choose a maxconn value. But maybe we should consider that instead of detecting that we're running with the large ulimit, we could as well detect that the maxconn value resulting from a large fdlimit seems very high **and** that the limit was not set by the user. It's a bit more complicated, but might work as well.
